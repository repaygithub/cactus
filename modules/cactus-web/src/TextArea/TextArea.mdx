---
name: TextArea
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import TextArea from './TextArea'
import cactusTheme from '@repay/cactus-theme'

# TextArea

## Best practices

- Use the `aria-labelledby` prop to associate labels with your text area to make them accessible.

## Basic usage

```jsx
import React, { useState, useCallback } from 'react'
import { Button, TextArea, Label } from '@repay/cactus-web'

export default function Form() {
  const [values, setValues] = useState({ textArea: '' })
  const handleChange = useCallback(
    event => {
      setValues({ ...values, [event.target.name]: event.target.value })
    }
  )
  const handleSubmit = useCallback(
    event => {
      event.preventDefault()
      const data = { ...values }
      // send data to api
    }
  )
  return (
    <form onSubmit={handleSubmit}>
      <Label id="label-id">Enter some text</Label>
      <TextArea aria-labelledby="label-id" name="textArea" onChange={handleChange} />
      <Button type="submit" variant="action">
        Submit
      </Button>
    </form>
  )
}
```

## Properties

<PropsTable of={TextArea} />
