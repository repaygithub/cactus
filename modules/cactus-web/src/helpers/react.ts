import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
