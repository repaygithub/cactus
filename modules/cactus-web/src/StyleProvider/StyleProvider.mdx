---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
