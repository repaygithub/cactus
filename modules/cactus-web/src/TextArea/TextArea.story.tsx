import { actions } from '@storybook/addon-actions'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'

import TextArea, { Status } from './TextArea'

type StatusOptions = { [k in Status | 'none']: Status | null }
const eventLoggers = actions('onChange', 'onFocus', 'onBlur')

const statusOptions: StatusOptions = {
  none: null,
  success: 'success',
  warning: 'warning',
  error: 'error',
}

storiesOf('TextArea', module).add(
  'Basic Usage',
  (): React.ReactElement => (
    <TextArea
      disabled={boolean('disabled', false)}
      placeholder={text('placeholder', 'Placeholder')}
      status={select('status', statusOptions, statusOptions.none)}
      resize={boolean('resize', false)}
      {...eventLoggers}
    />
  )
)
