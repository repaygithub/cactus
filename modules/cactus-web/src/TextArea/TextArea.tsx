import { BorderSize, CactusTheme, Shape, TextStyle } from '@repay/cactus-theme'
import PropTypes from 'prop-types'
import React from 'react'
import styled, {
  css,
  FlattenInterpolation,
  FlattenSimpleInterpolation,
  ThemeProps,
} from 'styled-components'
import { margin, MarginProps } from 'styled-system'

import { omitMargins } from '../helpers/omit'
import { textStyle } from '../helpers/theme'
import { Status, StatusPropType } from '../StatusMessage/StatusMessage'

export interface TextAreaProps
  extends React.TextareaHTMLAttributes<HTMLTextAreaElement>,
    MarginProps {
  disabled?: boolean
  status?: Status | null
  width?: string
  height?: string
  resize?: boolean
}

type StatusMap = { [K in Status]: FlattenInterpolation<ThemeProps<CactusTheme>> }

const statusMap: StatusMap = {
  success: css`
    border-color: ${(p): string => p.theme.colors.success};
    background: ${(p): string => p.theme.colors.transparentSuccess};
  `,
  warning: css`
    border-color: ${(p): string => p.theme.colors.warning};
    background: ${(p): string => p.theme.colors.transparentWarning};
  `,
  error: css`
    border-color: ${(p): string => p.theme.colors.error};
    background: ${(p): string => p.theme.colors.transparentError};
  `,
}

const displayStatus = (
  props: TextAreaProps
): FlattenInterpolation<ThemeProps<CactusTheme>> | string => {
  if (props.status && !props.disabled) {
    return statusMap[props.status]
  } else {
    return ''
  }
}

const borderMap: { [K in BorderSize]: ReturnType<typeof css> } = {
  thin: css`
    border: 1px solid;
  `,
  thick: css`
    border: 2px solid;
  `,
}

const shapeMap: { [K in Shape]: ReturnType<typeof css> } = {
  square: css`
    border-radius: 1px;
  `,
  intermediate: css`
    border-radius: 4px;
  `,
  round: css`
    border-radius: 8px;
  `,
}

const getBorder = (borderSize: BorderSize): FlattenInterpolation<ThemeProps<CactusTheme>> =>
  borderMap[borderSize]
const getShape = (shape: Shape): FlattenInterpolation<ThemeProps<CactusTheme>> => shapeMap[shape]

const Area = styled.textarea<TextAreaProps>`
  ${(p): FlattenInterpolation<ThemeProps<CactusTheme>> => getBorder(p.theme.border)}
  border-color: ${(p): string =>
    p.disabled ? p.theme.colors.lightGray : p.theme.colors.darkContrast};
  ${(p): FlattenInterpolation<ThemeProps<CactusTheme>> => getShape(p.theme.shape)}
  min-height: 100px;
  ${(p): string => (p.theme.mediaQueries ? p.theme.mediaQueries.small : '')}{
    min-width: 336px;

  }
  box-sizing: border-box;
  ${(p): FlattenSimpleInterpolation | TextStyle => textStyle(p.theme, 'body')}
  padding: 8px 16px;
  outline: none;
  background-color: ${(p): string =>
    p.disabled ? p.theme.colors.lightGray : p.theme.colors.white};
  height: ${(p): string => p.height || 'auto'};
  width: ${(p): string => p.width || 'auto'};
  display: block;
  resize: ${(p): string => (p.resize ? 'vertical' : 'none')};
  &:first-line {
    padding-right: 15px;
  }

  &:focus {
    border-color: ${(p): string => p.theme.colors.callToAction};
  }

  &::placeholder {
    color: ${(p): string => p.theme.colors.mediumContrast};
    font-style: oblique;
  }

  ${displayStatus}
`

const TextAreaBase = (props: TextAreaProps): React.ReactElement => {
  const { className, ...rest } = omitMargins(props)

  return (
    <div className={className}>
      <Area {...rest} />
    </div>
  )
}

export const TextArea = styled(TextAreaBase)`
  position: relative;
  display: block;

  ${margin}
`

// @ts-ignore
TextArea.propTypes = {
  disabled: PropTypes.bool,
  status: StatusPropType,
  width: PropTypes.string,
  height: PropTypes.string,
  resize: PropTypes.bool,
}

TextArea.defaultProps = {
  disabled: false,
  status: null,
  resize: false,
}

export default TextArea
