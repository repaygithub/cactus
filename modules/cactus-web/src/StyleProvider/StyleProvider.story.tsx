import React from 'react'

import { storiesOf } from '@storybook/react'
import Box from '../Box/Box'
import cactusTheme from '@repay/cactus-theme'
import Grid from '../Grid/Grid'
import styled from 'styled-components'
import StyleProvider from './StyleProvider'

const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const BreakpointBox = styled(Box)`
  max-width: 320px;
  width: 50px;
  height: 50px;
  border-radius: 15%

  ${p => p.theme.mediaQueries.small} {
    max-width: 708px;
    width: 75px;
    height 75px;
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    max-width: 964px;
    width: 100px;
    height: 100px;
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    max-width: 1140px;
    width: 150px;
    height: 150px;
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extralarge} {
    max-width: 1380px;
    width: 200px;
    height: 200px;
    border-radius: 50%;
  }
`

storiesOf('Style Provider', module).add('Component Adjusments based on Media Queries', () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <Grid justify="center">
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
      </Grid>
    </StyleProvider>
  )
})
