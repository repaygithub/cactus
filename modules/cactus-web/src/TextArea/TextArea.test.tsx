import * as React from 'react'

import { cleanup, render } from '@testing-library/react'
import { generateTheme } from '@repay/cactus-theme'
import { StyleProvider } from '../StyleProvider/StyleProvider'
import TextArea from './TextArea'
import userEvent from '@testing-library/user-event'

afterEach(cleanup)

describe('component: TextArea', () => {
  test('should render a textarea', () => {
    const { container } = render(
      <StyleProvider>
        <TextArea />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render a disabled textarea', () => {
    const { container } = render(
      <StyleProvider>
        <TextArea disabled={true} />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render a textarea with a placeholder', () => {
    const { container } = render(
      <StyleProvider>
        <TextArea placeholder="Type something!" />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render a success textarea', () => {
    const { container } = render(
      <StyleProvider>
        <TextArea status="success" />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render a warning textarea', () => {
    const { container } = render(
      <StyleProvider>
        <TextArea status="warning" />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render an error textarea', () => {
    const { container } = render(
      <StyleProvider>
        <TextArea status="error" />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should support margin space props', () => {
    const { container } = render(
      <StyleProvider>
        <TextArea ml={5} />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should trigger onChange handler', () => {
    const onChange = jest.fn()
    const { getByPlaceholderText } = render(
      <StyleProvider>
        <TextArea onChange={onChange} placeholder="get me" />
      </StyleProvider>
    )

    userEvent.type(getByPlaceholderText('get me'), 'typing in a textarea')
    expect(onChange).toHaveBeenCalled()
  })

  describe('with theme customization', () => {
    test('should have 1px border', () => {
      const theme = generateTheme({ primaryHue: 200, border: 'thin' })
      const { asFragment } = render(
        <StyleProvider theme={theme}>
          <TextArea />
        </StyleProvider>
      )

      expect(asFragment()).toMatchSnapshot()
    })

    test('should have 8px border radius', () => {
      const theme = generateTheme({ primaryHue: 200, shape: 'intermediate' })
      const { asFragment } = render(
        <StyleProvider theme={theme}>
          <TextArea />
        </StyleProvider>
      )

      expect(asFragment()).toMatchSnapshot()
    })

    test('should have 1px border radius', () => {
      const theme = generateTheme({ primaryHue: 200, shape: 'square' })
      const { asFragment } = render(
        <StyleProvider theme={theme}>
          <TextArea />
        </StyleProvider>
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
