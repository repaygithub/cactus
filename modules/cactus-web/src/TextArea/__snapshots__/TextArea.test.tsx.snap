// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: TextArea should render a disabled textarea 1`] = `
.c1 {
  border: 1px solid;
  border-color: hsl(0,0%,90%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,90%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1"
      disabled=""
    />
  </div>
</div>
`;

exports[`component: TextArea should render a success textarea 1`] = `
.c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
  border-color: hsl(145,89%,28%);
  background: hsla(145,89%,28%,0.3);
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</div>
`;

exports[`component: TextArea should render a textarea 1`] = `
.c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</div>
`;

exports[`component: TextArea should render a textarea with a placeholder 1`] = `
.c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1"
      placeholder="Type something!"
    />
  </div>
</div>
`;

exports[`component: TextArea should render a warning textarea 1`] = `
.c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
  border-color: hsl(47,82%,47%);
  background: hsla(47,82%,47%,0.3);
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</div>
`;

exports[`component: TextArea should render an error textarea 1`] = `
.c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
  border-color: hsl(353,84%,44%);
  background: hsla(353,84%,44%,0.3);
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</div>
`;

exports[`component: TextArea should support margin space props 1`] = `
.c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
  margin-left: 24px;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</div>
`;

exports[`component: TextArea with theme customization should have 1px border radius 1`] = `
<DocumentFragment>
  .c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 1px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</DocumentFragment>
`;

exports[`component: TextArea with theme customization should have 2px border 1`] = `
<DocumentFragment>
  .c1 {
  border: 2px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 8px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</DocumentFragment>
`;

exports[`component: TextArea with theme customization should have 8px border radius 1`] = `
<DocumentFragment>
  .c1 {
  border: 1px solid;
  border-color: hsl(200,9%,35%);
  border-radius: 4px;
  min-height: 100px;
  min-width: 336px;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  padding: 8px 16px;
  outline: none;
  background-color: hsl(0,0%,100%);
  height: auto;
  width: auto;
  display: block;
  resize: none;
}

.c1:first-line {
  padding-right: 15px;
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: block;
}

<div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</DocumentFragment>
`;
